### ISSUE TITLE
[Bash] Heredoc with compact-command is not scoped correctly

### ISSUE THREAD

**Issue by jfcherng** [sublimehq/Packages#1190]
Just tried the new re-written Bash syntax and found this issue.

```bash
print_info_text_compact () { cat <<EOF; do_something; }
This is some info text.
EOF

# <- scope here should be only "source.shell.bash"
```

![image](https://user-images.githubusercontent.com/6594915/31256377-ac6d67a4-aa64-11e7-8664-32512341541e.png)

**Comment by rwols** [comment]
This is an example where the heredoc trick is lacking...

So, what is happening is this:

1. We match the start of a function `print_info_text_compact () {`, scope it appropriately. Context stack looks like this now:
   ```
   function-definition
   main
   ```
2. We encounter the start of a heredoc-token. Context stack looks like this now:
   ```
   heredoc-preamble
   heredoc-body
   function-definition
   main
   ```
3. We reach the end of the line. This pops off the `heredoc-preamble` context:
   ```
   heredoc-body
   function-definition
   main
   ```
4. We encounter the heredoc (`EOF`) token that we matched before, this pops off the `heredoc-body` context:
   ```
   function-definition
   main
   ```
Essentially what *should* happen is that in step (3), if we find a `}`, we should look into the context stack and remove the first `function-definition` context that we find:

```
   heredoc-preamble
   heredoc-body
   function-definition <--- cannot reach :(
   main
```

**Comment by deathaxe** [comment]
I guess there is no way to handle HEREDOCs properly without loosing all the `meta.block` etc. stuff. We would need to not push into any context. We'll always find situations in which herdocs fail.

**Comment by rwols** [comment]
Yes. There's also the issue of multiple heredocs starting on a single line.

```bash
#!/bin/bash
cat /dev/fd/3 3<< FIRST /dev/fd/4 4<< SECOND
hello there
FIRST
general kenobi
SECOND
```
Prints
```
hello there
general kenobi
[Finished in 10ms]
```
